src/msg_recorder/recorder.cc

```diff
@@ -21,13 +21,22 @@
 
 namespace cris::core {
 
-MessageRecorder::MessageRecorder(const RecorderConfig& recorder_config, std::shared_ptr<JobRunner> runner)
+MessageRecorder::MessageRecorder(RecorderConfig recorder_config, std::shared_ptr<JobRunner> runner)
     : Base(std::move(runner))
-    , record_dir_(recorder_config.record_dir_ / RecordDirNameGenerator())
+    , recorder_config_(std::move(recorder_config))
```

Comment:
为什么要move？直接在这里复制不就好了，何必在参数上复制一份？

Reply:
rval是匹配const &的，没问题，也不会有多余的复制

Reply:
如果编译器优化做完整，对于f(rval)的结果是rval直接被写到目标const &中（或者说共享同一片存储），然后目标构造被一个copy elimination处理掉。

Reply:
(pending issue)

Reply:
是的，所以不需要move，而是走copy elision的
https://en.cppreference.com/w/cpp/language/copy_elision

> Under the following circumstances, the compilers are required to omit the copy and move construction of class objects, even if the copy/move constructor and the destructor have observable side-effects.
> ...
> In the initialization of an object, when the initializer expression is a prvalue of the same class type (ignoring cv-qualification) as the variable type:

Reply:
想了一下我觉得你说的是对的，这确实是一种可行的技巧。
但还有一个问题，我们希望能const的地方尽量const，保证安全性和可读性，避免意外修改的同时告诉读者这个变量不会被修改，这样编译器可以帮我们做很多static analysis的活。
- 在这个不涉及性能的地方，我们是否有必要牺牲这些优点来换取可能减少的一次copy（其实如果只是没有side effect的默认复制，编译器有data propagation，最后大概率还是会优化掉）。
- 在涉及性能的地方，我觉得用模版更合适，还能保留perfect forwarding。

Reply:
> 被move的是一个临时对象（parameter），没问题啊

被move的是一个lvalue，不临时，它在生存期覆盖了整个函数，不const就意味着它可能在函数里被改掉，编译器不会帮我们阻止这件事。

> 我们代码库里这么用是很常见的，下面就是其中两例：

job queue那个我觉得没毛病，因为那块性能很重要，无论用什么trick提升性能都值得。
node那个我觉得没必要。

@chenhao94 你觉得呢？这边代码之前应该是你写的？

Reply:
> 在初始化列表被move了，在函数体里被修改也无妨

为什么无妨？
如果它被修改（包括被move），在body里再次引用的时候，有可能还以为是原来传入的值，这算一种常见的error pattern吧？

Reply:
> 避免用错是程序员要做的

避免任何bug都是程序员要做的，但程序员还是天天写bug，所以才需要工程上的技巧来弥补人的缺陷嘛。

> 我们争论的根本在于：这里性能是否重要

我的理解是：
- 这个函数是不会被高频调用的，所以性能不重要
- 减少copy并不能改善性能，因为复杂度还是一样的，初始化1个长度为n的vector和初始化并复制它，都是O(n)的

如果是会影响性能的地方，我完全赞同你的处理。

Reply:
@xkszltl 我觉得这里没有问题，而且在需要存下参数的对象时是应当鼓励的写法，因为
- 如果参数是const&会强制复制，不合适
- 如果参数是&&，而caller希望保留自己的copy，则必须显式在caller侧做复制，啰嗦
- 如果参数是传值，那么caller可以自行选择是否复制，代价只是可能会多一次move ctor
- perfect forwarding必须写成模板，这里不用

这种写法的问题是
- caller可能会无意间复制。这其实和优点的第二条是同一件事情。就看你是在意灵活性还是想让caller更不容易复制
- 像你说的，函数里会move掉参数，所以参数不可以是const，也有use-after-move的可能。但这一点我觉得其实不是个问题，因为任何用move的地方都是一样的（而且参数用&&也是一样的）。

这两点我没记错的话，clang-tidy都是可以识别的，所以也没那么依赖reviewer

另外一个问题是
- caller给rvalue参数的时候，多一次move ctor

可是move ctor一般被认为是很快的。如果不是特别在意性能（在意到一个move ctor都嫌多）的地方，这么写都没有问题。

Reply:
我的关注点主要还是在性能上，觉得不影响性能不需要抠常数的地方整这些不值得，因为吃不到好处，只剩代价了。
吃不到好处的意思是，如果外面给进来的是rvalue，dataflow只可能是链状（rvalue不可在外面复用），所以copy complexity = O(init complexity of rvalue)，是否copy复杂度都一样。

> 这两点我没记错的话，clang-tidy都是可以识别的，所以也没那么依赖reviewer

That's a good point.
If it does, it would make sense to allow this specific pattern of "non-const lvalue in ctor + immediate move it init list".

BTW please don't replace copy/move ctor with this.

Reply:
> 我的关注点主要还是在性能上，觉得不影响性能不需要抠常数的地方整这些不值得，因为吃不到好处，只剩代价了。

正是因为没必要弄复杂，才直接传值，而不去搞什么&&或者perfect forwarding。const&也可以，不过就是没法const和一定有一次多余复制二选一。我选前面的，因为我觉得没有const不算什么问题。